V1.0.0 RC (#182)
OWASP/maswe: weaknesses/MASVS-STORAGE/MASWE-0003.md

Reviewer: Stefan: DONE

mappings:
  masvs-v1: [MSTG-STORAGE-2]
  masvs-v2: [MASVS-STORAGE-1, MASVS-STORAGE-2, MASVS-CRYPTO-2, MASVS-PRIVACY-1]
  cwe: [200, 284, 312, 313, 732, 922]
  android-risks:
  - file-providers
  maswe-beta: [MASWE-0006, MASWE-0002, MASWE-0118]
refs:
- https://developer.apple.com/documentation/uikit/protecting_the_user_s_privacy/encrypting_your_app_s_files
- https://developer.android.com/about/versions/nougat/android-7.0-changes#permfilesys
- https://developer.android.com/privacy-and-security/security-tips#internal-storage
status: new
---

## Overview

This weakness occurs when an app stores sensitive data unencrypted in private storage locations, such as the application sandbox, where it can be exposed via incorrect file permissions, an app or device vulnerability, or data backup mechanisms.

Sensitive data may include personally identifiable information (PII), passwords, cryptographic keys, or session tokens.

## Modes of Introduction

- **Data Stored Unencrypted**: Writing sensitive data to the app's private data directory (sandbox) unencrypted.
- **Hardcoded Encryption Key**: Encrypting sensitive data with a key that is hardcoded inside the application.

> - **Encryption Key Stored on Filesystem**: Encrypting sensitive data but storing the key alongside it or in another easily accessible location.

Reviewer: ```suggestion
- **Encryption Key Stored on Filesystem**: Encrypting sensitive data but storing the generated key alongside it or in another easily accessible location.
```

- **Insufficient Encryption**: Encrypting sensitive data with an algorithm or configuration that is not considered strong.

> - **Insufficient Access Restrictions**: Exposing private files to other apps through incorrect file permissions (e.g. the deprecated `MODE_WORLD_READABLE`/`MODE_WORLD_WRITEABLE` modes on Android), a misconfigured `FileProvider`, or Keychain items protected with weak accessibility attributes on iOS (e.g. `kSecAttrAccessibleAlways`).

Reviewer: ```suggestion
- **Insufficient Access Restrictions**: Exposing private files to other apps through incorrect file permissions (e.g. the deprecated `MODE_WORLD_READABLE`/`MODE_WORLD_WRITEABLE` file permission modes on Android), a misconfigured `FileProvider`, or Keychain items protected with weak accessibility attributes on iOS (e.g. `kSecAttrAccessibleAlways`).
```

- **Data Not Removed After Use**: Retaining sensitive data in private storage (including caches, temporary files, WebView state, and network caches) longer than needed.

## Impact

- **Compromise of Sensitive Data**: Attackers can extract PII and other user data from the application sandbox, resulting in unauthorized disclosure and enabling further attacks such as identity theft.
- **Authentication or Authorization Bypass**: Attackers can extract passwords, cryptographic keys, or session tokens, resulting in account takeover or unauthorized access to protected functionality.

## Mitigations

- **Minimize Local Storage of Sensitive Data**: Avoid storing sensitive data locally if it is not required for app functionality, e.g. keep PII server-side, render it at time of use, and remove any cached data on logout.

> - **Restrict File Access**: Never use `MODE_WORLD_READABLE`/`MODE_WORLD_WRITEABLE`; share data through a properly configured `ContentProvider`/`FileProvider` with per-grant permissions instead, and on iOS store Keychain items with the strictest viable accessibility attribute (e.g. `kSecAttrAccessibleWhenUnlockedThisDeviceOnly`).

Reviewer: ```suggestion
- **Restrict File Access**: Use platform APIs to restrict file access as much as possible (least privilege principle). On Android, share data through a properly configured `ContentProvider`/ `FileProvider` with per-grant permissions instead of the deprecated `MODE_WORLD_READABLE`/`MODE_WORLD_WRITEABLE` file permission modes. On iOS, store Keychain items with the strictest viable accessibility attribute (e.g. `kSecAttrAccessibleWhenUnlockedThisDeviceOnly`).
```

Remark: I rephrased the beginning to be more generic (use the platform API).
